service/r2/routes.go

```diff
@@ -271,3 +275,117 @@ func fetchRollupRuleHistory(s *service, r *http.Request) (data interface{}, err
 	}
 	return models.NewRollupRuleSnapshots(hist), nil
 }
+
+func bulkUpdateRuleSet(s *service, r *http.Request) (data interface{}, err error) {
+	vars := mux.Vars(r)
+	var req bulkRequest
+	if err := parseRequest(&req, r.Body); err != nil {
+		return nil, err
+	}
+	uOpts, err := s.newUpdateOptions(r)
+	if err != nil {
+		return nil, err
+	}
+
+	originalRS, err := s.store.FetchRuleSet(vars[namespaceIDVar])
+	if err != nil {
+		return nil, err
+	}
+	if req.RuleSetVersion != originalRS.Version() {
+		return nil, NewConflictError(
+			fmt.Sprintf("can not update namespace %s, version mismatch", vars[namespaceIDVar]),
+		)
+	}
+
+	newRS, err := applyChangesToRuleSet(
+		req.RuleSetChanges,
+		originalRS.ToMutableRuleSet().Clone(),
+		uOpts,
+		s.updateHelper,
+	)
+	if err != nil {
+		return nil, NewBadInputError(err.Error())
+	}
+
+	rs, err := s.store.UpdateRuleSet(newRS)
+	if err != nil {
+		return nil, err
+	}
+	latest, err := rs.Latest()
+	if err != nil {
+		return nil, err
+	}
+	serializableRuleSet := models.NewRuleSet(latest)
+
+	return serializableRuleSet, nil
+}
+
+func applyChangesToRuleSet(
```

Comment:
As mentioned previously, it might be worth taking a RuleSetSnapshot and then doing the .ToMutable().Clone() inside here.